mempool/reactor.go

```diff
@@ -262,21 +264,72 @@ func (memR *Reactor) broadcastTxRoutine(peer p2p.Peer) {
 	}
 }
 
+// txs iterates over the transaction list and builds a batch of txs. next is
+// included.
+// WARNING: mutates next!
+func (memR *Reactor) txs(next *clist.CElement, peerID uint16, peerHeight int64) [][]byte {
+	batch := make([][]byte, 0, memR.config.TxBatchSize)
+
+	memTx := next.Value.(*mempoolTx)
+	if _, ok := memTx.senders.Load(peerID); !ok {
+		batch = append(batch, []byte(memTx.tx))
+	}
+
+	if len(batch) >= memR.config.TxBatchSize {
+		return batch
+	}
+
+	n := next.Next()
+	for n != nil {
+		memTx = n.Value.(*mempoolTx)
+		if peerHeight < memTx.Height()-1 {
+			break
+		}
+
+		if _, ok := memTx.senders.Load(peerID); !ok {
+			batch = append(batch, []byte(memTx.tx))
+		}
+
+		// Advance next (non-nil) pointer, which is used later to
+		// wait (NextWaitChan) for next transaction to appear.
+		next = n
+		// Advance n (could be nil) pointer
+		n = n.Next()
+
+		if len(batch) >= memR.config.TxBatchSize {
+			return batch
+		}
+	}
+
+	return batch
+}
+
 //-----------------------------------------------------------------------------
 // Messages
 
-func (memR *Reactor) decodeMsg(bz []byte) (TxMessage, error) {
+func (memR *Reactor) decodeMsg(bz []byte) (TxsMessage, error) {
 	msg := protomem.Message{}
 	err := msg.Unmarshal(bz)
 	if err != nil {
-		return TxMessage{}, err
+		return TxsMessage{}, err
 	}
 
-	var message TxMessage
+	var message TxsMessage
+
+	if i, ok := msg.Sum.(*protomem.Message_Txs); ok {
+		txs := i.Txs.GetTxs()
```

Comment:
```suggestion
	if msg := msg.GetTxs(); msg != nil {
		txs := msg.GetTxs()
```

Reply:
have a slight preference over my version since it asserts type, not method of one of the messages (GetTxs)